### PR TITLE
Add hybrid jobs to Braket QAOA demo

### PR DESCRIPTION
**Pull request by mbeach-aws**
**Title:**
Computing gradients in parallel with Amazon Braket

**Summary:**
Adding hybrid jobs to Braket QAOA demo. 

**Relevant references:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

**Review by josh146 — COMMENTED** [review]
Thanks @mbeach-aws! Don't forget to also include the metadata (might need to update the github notebook link?)

File: demonstrations/braket-parallel-gradients.py [hunk 1]
```diff
@@ -441,9 +435,160 @@ def circuit(params, **kwargs):
...
+# such as limited memory. With Amazon Braket Hybrid Jobs, you can send your entire algorithm, both
+# classical and quantum parts, to run on AWS while you get a coffee!
+#
+# With a single-line-of-code, we'll see how to scale from PennyLane simulators on your laptop to
+# running full-scale experiments on AWS that leverage both powerful classical compute and quantum
+# devices. For more details on hybrid jobs see [LINK TO PENNYLANE DEMO 1].
+# .. warning::
+#     The following demo is only compatible with Python version 3.10.
+#
+
```
Comment (KetpuntoG) [inline demonstrations/braket-parallel-gradients.py:454]:
```suggestion
# devices. For more details on hybrid jobs see [LINK TO PENNYLANE DEMO 1].
#
# .. warning::
#     The following demo is only compatible with Python version 3.10.
#

```

File: demonstrations/braket-parallel-gradients.py [hunk 2]
```diff
@@ -441,9 +435,160 @@ def circuit(params, **kwargs):
...
+from braket.jobs.decorator import hybrid_job
```
Comment (KetpuntoG) [inline demonstrations/braket-parallel-gradients.py:455]: I lacked a bit of context to introduce this code. It is a large cell and I would like to know what is the idea before what we see it. Something like:
In the following cell we will encapsulate the whole QAOA process in the same function to get [..........]

File: demonstrations/braket-parallel-gradients.py [hunk 3]
```diff
@@ -441,9 +435,160 @@ def circuit(params, **kwargs):
...
+#                       [0.3853200394389563, 0.2402391372931216]],
+#        'final_cost': -37.76176577995996,
+#        'braket_tasks_cost': 0}
+
+
+##############################################################################
+# The results included the three values from the return statement of our function.
+#
+# Additionally, we can retrieve the metrics recorded during the training with:
+#
```
Comment (KetpuntoG) [inline demonstrations/braket-parallel-gradients.py:553]:
```suggestion
# The results included the three values from the return statement of our function.
# Additionally, we can retrieve the metrics recorded during the training with:
#
```

File: demonstrations/braket-parallel-gradients.py [hunk 4]
```diff
@@ -441,9 +435,160 @@ def circuit(params, **kwargs):
...
+
+##############################################################################
+# The metrics are plotted below.
+# Plotting the convergence of the loss function metric
+#
+
+import pandas as pd
+import matplotlib.pyplot as plt
+
+metrics = job.metrics()
```
Comment (KetpuntoG) [inline demonstrations/braket-parallel-gradients.py:565]: repeated metrics = job.metrics(), could you rephrase this part not to repeat?

File: demonstrations/braket-parallel-gradients.py [hunk 5]
```diff
@@ -441,9 +435,160 @@ def circuit(params, **kwargs):
...
+# .. figure:: ../_static/qaoa_training.png
+#     :align: center
+#     :scale: 75%
+#     :alt: Convergence of cost function for QAOA training.
+#     :target: javascript:void(0);
+
+##############################################################################
+# Great! The algorithm converged as expected.
+#
+# For more examples of how to use hybrid jobs, see the
```
Comment (KetpuntoG) [inline demonstrations/braket-parallel-gradients.py:583]: Even if there wasn't in this one, could you write a slightly longer conclusion following the format of other more [recent demos](https://pennylane.ai/qml/demos/tutorial_optimal_control#conclusion)? (It doesn't have to be that long either, but a final thought is always good 😄)

File: demonstrations/braket-parallel-gradients.py [hunk 6]
```diff
@@ -441,9 +435,160 @@ def circuit(params, **kwargs):
...
+# We have seen how we can use PennyLane on Braket to solve graph optimization problems with QAOA.
+# However, we have only used Amazon Braket for the quantum simulations, leaving the classical
+# Python code to run on our laptop.
+# This is great for getting started with small experiments, but for large-scale algorithms running
+# them on a laptop is impractical due to potential crashes, restarts or even hardware limitations
+# such as limited memory. With Amazon Braket Hybrid Jobs, you can send your entire algorithm, both
+# classical and quantum parts, to run on AWS while you get a coffee!
+#
+# With a single-line-of-code, we'll see how to scale from PennyLane simulators on your laptop to
+# running full-scale experiments on AWS that leverage both powerful classical compute and quantum
```
Comment (KetpuntoG) [inline demonstrations/braket-parallel-gradients.py:449]:
```suggestion
# run full-scale experiments on AWS that leverage both powerful classical compute and quantum
```
not super sure, is this correct?

File: demonstrations/braket-parallel-gradients.py [hunk 7]
```diff
@@ -441,9 +435,160 @@ def circuit(params, **kwargs):
...
+#       'QUEUED'
+
+##############################################################################
+# The hybrid job will be scheduled to run and will appear in the "QUEUED" state.
+# If the target is a QPU, the job will be queued with other jobs.
+# If the target device is not a QPU, the hybrid job should start immediately.
+# Note that since the algorithm code is run in a containerized environment, it takes approximately 1
+# minute to start running your algorithm.
 #
+# After the hybrid job completes, we can get the results with ``job.result()``. For this example, it
```
Comment (KetpuntoG) [inline demonstrations/braket-parallel-gradients.py:532]:
```suggestion
# After the hybrid job is completed, we can get the results with ``job.result()``. For this example, it
```

**Review by KetpuntoG — COMMENTED** [review]
Good addition! 🚀 Apart from the comments, could you also add in the opening paragraph that hybrid jobs will be used in the demo?

File: demonstrations/braket-parallel-gradients.py [hunk 8]
```diff
@@ -441,9 +435,160 @@ def circuit(params, **kwargs):
...
+metrics = job.metrics()
+
+##############################################################################
+# The metrics are plotted below.
+# Plotting the convergence of the loss function metric
+#
+
+import pandas as pd
+import matplotlib.pyplot as plt
+
```
Comment (KetpuntoG) [inline demonstrations/braket-parallel-gradients.py:564]: could you use our plt style format?

`plt.style.use('pennylane.drawer.plot')
`

In this case, you would have to regenerate the image

**Comment by KetpuntoG** [comment]
Hi @mbeach-aws , don't forget to work on the last proposed suggestions 😄

File: demonstrations/braket-parallel-gradients.py [hunk 9]
```diff
@@ -437,9 +437,173 @@ def circuit(params, **kwargs):
...
+# run full-scale experiments on AWS that leverage both powerful classical compute and quantum
+# devices. For more details on hybrid jobs see
+# :doc:`getting started with hybrid jobs <getting_started_with_hybrid_jobs>`.
+#
+# .. warning::
+#     The following demo is only compatible with Python version 3.10.
+#
+# In the following code, we run the same QAOA optimization as before, but this time on an
+# `Amazon EC2 ml.c5.xlarge instance <https://aws.amazon.com/ec2/instance-types/c5/>`__
+# instead of our laptop. We specify this with the `instance_config` for our hybrid job.
```
Comment (KetpuntoG) [inline demonstrations/braket-parallel-gradients.py:460]:
```suggestion
# instead of our laptop. We specify this with the `InstanceConfig` for our hybrid job.
```

Would it be more accurate?

File: demonstrations/braket-parallel-gradients.py [hunk 10]
```diff
@@ -437,9 +437,173 @@ def circuit(params, **kwargs):
...
+# Conclusion
+# ----------------------------------
+# In this tutorial, we found the solution to a MaxCut problem using QAOA. We used Amazon Braket
+# on-demand simulators to efficiently parallelize the parameter-shift rule for gradients. We also
+# demonstrated how to scale-up experiments with your own simulator, such as PennyLane Lightning, by
+# running on Amazon Braket Hybrid Jobs. In both approaches, we perform the intensive calculations on
+# the cloud instead of our laptop.
+#
+# For more examples of how to use hybrid jobs, see the
+# `Amazon Braket examples GitHub <https://github.com/amazon-braket/amazon-braket-examples/tree/main/examples/hybrid_jobs>`__
```
Comment (KetpuntoG) [inline demonstrations/braket-parallel-gradients.py:599]:
```suggestion
# `Amazon Braket examples GitHub <https://github.com/amazon-braket/amazon-braket-examples/tree/main/examples/hybrid_jobs>`__.
```

File: demonstrations/braket-parallel-gradients.py [hunk 11]
```diff
@@ -437,9 +437,173 @@ def circuit(params, **kwargs):
...
+##############################################################################
+# The results included the three values from the return statement of our function.
+# Additionally, we can retrieve the metrics recorded during the training with:
+#
+
+metrics = job.metrics()
+
+##############################################################################
+# The metrics are plotted below.
+# Plotting the convergence of the loss function metric
```
Comment (DanielNino27) [inline demonstrations/braket-parallel-gradients.py:566]: This sentence feels more like commenting on the code. as it is not a complete sentence. I would suggest rephrasing it so that it is a complete sentence or putting it in the code as a comment.

**Review by DanielNino27 — COMMENTED** [review]
One minor comment - the rest looks good to me.

**Review by KetpuntoG — APPROVED** [review]
Great to see this! Well done 🚀

**Review by DanielNino27 — APPROVED** [review]
